include/picongpu/particles/collision/InterCollision.hpp

```diff
@@ -0,0 +1,358 @@
+             * @tparam T_FilterPair A pair of particle filters, each for each species
+             *     in the colliding pair.
+             * @tparam T_Species0 1st colliding species.
+             * @tparam T_Species1 2nd colliding species.
+             */
+            template<
+                typename T_CollisionFunctor,
+                typename T_Params,
+                typename T_FilterPair,
+                typename T_Species0,
+                typename T_Species1>
+            struct DoInterCollision
+            {
+                /* Run kernel
+                 *
+                 * @param deviceHeap A pointer to device heap for allocating particle lists.
+                 * @param currentStep The current simulation step.
+                 */
+                HINLINE void operator()(const std::shared_ptr<DeviceHeap>& deviceHeap, uint32_t currentStep)
+                {
+                    using Species0 = T_Species0;
+                    using FrameType0 = typename Species0::FrameType;
+                    using Filter0 = typename T_FilterPair::first ::template apply<Species0>::type;
+
+                    using Species1 = T_Species1;
+                    using FrameType1 = typename Species1::FrameType;
+                    using Filter1 = typename T_FilterPair::second ::template apply<Species1>::type;
+
+                    using CollisionFunctor = T_CollisionFunctor;
+
+                    // Access particle data:
+                    DataConnector& dc = Environment<>::get().DataConnector();
+                    auto species0 = dc.get<Species0>(FrameType0::getName(), true);
+                    auto species1 = dc.get<Species1>(FrameType1::getName(), true);
+
+                    // Use mapping information from the first species:
+                    AreaMapping<CORE + BORDER, picongpu::MappingDesc> mapper(species0->getCellDescription());
+
+                    constexpr uint32_t numWorkers
+                        = pmacc::traits::GetNumWorkers<pmacc::math::CT::volume<SuperCellSize>::type::value>::value;
+
+                    //! random number generator
+                    using RNGFactory = pmacc::random::RNGProvider<simDim, random::Generator>;
+                    constexpr float_X coulombLog = T_Params::coulombLog;
+
+                    PMACC_KERNEL(InterCollision<numWorkers>{})
+                    (mapper.getGridDim(), numWorkers)(
+                        species0->getDeviceParticlesBox(),
+                        species1->getDeviceParticlesBox(),
+                        mapper,
+                        deviceHeap->getAllocatorHandle(),
+                        RNGFactory::createHandle(),
+                        CollisionFunctor(currentStep),
+                        coulombLog,
+                        Filter0(),
+                        Filter1());
+
+                    // Release particle data:
+                    // TODO: can I relase it like this? Or do I have to synchronize, make KERNEL call blocking?
```

Comment:
Yes this is correct, currently `release` is doing nothing but is already added to the code for future use.